feat: detect duplicate members in struct definitions

EricHasegawa opened the pull request:
Detects duplicate arguments in struct definitions. Note that there's a simpler way to do this where we'd only catch errors of the `Struct.new(:foo, :foo)` variety (and not those involving strings), but we could remove the functionality where we handle strings in addition to symbols.

On this note - is there any specific reason _why_ we previously bailed on line 117 if it's not a symbol? Curious because I'm not 100% sure this change does not have implications beyond detecting duplicate args.


### Motivation
Fixes https://github.com/sorbet/sorbet/issues/8275


### Test plan
See included automated tests.

EricHasegawa commented on test/testdata/rewriter/struct_duplicate_args.rb:
```diff
@@ -0,0 +1,20 @@
+class TestStructs
+    DuplicateStruct1 = Struct.new(:foo, :foo) # error: Duplicate member 'foo' in Struct definition
+
+    DuplicateStruct2 = Struct.new(:foo, 'foo') # error: Duplicate member 'foo' in Struct definition
+  
+    DuplicateStruct3 = Struct.new("FooStructDup", :foo, :bar, :foo) # error: Duplicate member 'foo' in Struct definition
```
if we keep the check on line 117 where we early return if an arg is not a symbol, these tests will fail and we won't cover these errors.

froydnj commented on rewriter/Struct.cc:
```diff
@@ -112,14 +112,32 @@ vector<ast::ExpressionPtr> Struct::run(core::MutableContext ctx, ast::Assign *as
-        auto symLoc = sym->loc;
-        auto strname = name.shortName(ctx);
+
+        std::string strname;
+        if (lit->isSymbol()) {
+            strname = lit->asSymbol().show(ctx);
+        } else if (lit->isString()) {
+            strname = lit->asString().show(ctx);
+        } else {
+            return empty;
+        }
+
+        core::NameRef name = ctx.state.enterNameUTF8(strname);
```
There should be no need to do this conversion from `NameRef` to string and then back to `NameRef`: `asSymbol` and `asString` already return `NameRef`s, and a symbol `:foo` and a string `"foo"` are guaranteed to have the same underlying `NameRef`.  So you can use the `NameRef` of the literal directly:

```c++
core::NameRef name;
if (lit->isSymbol()) {
    name = lit->asSymbol();
} else if (lit->isString()) {
    name = lit->asString();
} else {
    return empty;
}
```

froydnj commented on rewriter/Struct.cc:
```diff
@@ -112,14 +112,32 @@ vector<ast::ExpressionPtr> Struct::run(core::MutableContext ctx, ast::Assign *as
+
+        core::NameRef name = ctx.state.enterNameUTF8(strname);
+
+        if (seenNames.find(name) != seenNames.end()) {
+            if (auto e = ctx.beginError(lit->loc, core::errors::Rewriter::InvalidStructMember)) {
+                e.setHeader("Duplicate member '{}' in Struct definition", name.show(ctx));
+            }
+            return empty;
+        }
+        seenNames.insert(name);
```
It would be better to do:

```c++
auto [_it, inserted] = seenNames.insert(name);
if (!inserted) {
    if (auto e = ctx.beginError(...)) {
        ...
    }
}
```

so that you're only looking up names in the set a single time.

froydnj reviewed (changes requested):
Thanks for the PR!  A few notes below.

froydnj commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_RqYFtwk6rfqIkp
→ https://go/builds/bui_RqYGQSkgJyFb44
→ https://go/builds/bui_RqYGpL5cxLR8yc

jez commented on rewriter/Struct.cc:
```diff
@@ -112,13 +112,30 @@ vector<ast::ExpressionPtr> Struct::run(core::MutableContext ctx, ast::Assign *as
-        auto sym = ast::cast_tree<ast::Literal>(arg);
-        if (!sym || !sym->isSymbol()) {
+        auto lit = ast::cast_tree<ast::Literal>(arg);
+        if (!lit) {
```
```suggestion
        if (!lit || !lit->isName()) {
```

jez commented on rewriter/Struct.cc:
```diff
@@ -112,13 +112,30 @@ vector<ast::ExpressionPtr> Struct::run(core::MutableContext ctx, ast::Assign *as
-        core::NameRef name = sym->asSymbol();
-        auto symLoc = sym->loc;
+
+        core::NameRef name;
+        if (lit->isSymbol()) {
+            name = lit->asSymbol();
+        } else if (lit->isString()) {
+            name = lit->asString();
+        } else {
+            return empty;
+        }
```
```suggestion
        auto name = lit->asName();
```

jez commented on rewriter/Struct.cc:
```diff
@@ -112,13 +112,30 @@ vector<ast::ExpressionPtr> Struct::run(core::MutableContext ctx, ast::Assign *as
+        auto [it, inserted] = seenNames.insert(name);
+        if (!inserted) {
+            if (auto e = ctx.beginError(lit->loc, core::errors::Rewriter::InvalidStructMember)) {
+                e.setHeader("Duplicate member '{}' in Struct definition", name.show(ctx));
```
```suggestion
                e.setHeader("Duplicate member `{}` in `{}` definition", name.show(ctx), "Struct");
```

Using the pattern

```
`{}`
```

will mean that that substring gets highlighted cyan in terminal output.

jez commented on rewriter/Struct.cc:
Most of this logic could likely also apply to the `rewriter/Data.cc` rewriter.

It might be worth factoring it into a helper in `rewriter/Util.{h,cc}` and using it in both places.

EricHasegawa replied:
Refactored it to handle both cases (and added tests for `Data.cc` changes), there's a slight performance tradeoff here since we run the duplicate arg check in  a separate note now, I don't feel strongly about doing them together or having the same logic twice so let me know if you'd like me to change it back or know of an alternate approach

jez commented:
> On this note - is there any specific reason _why_ we previously bailed on line 117 if it's not a symbol? Curious because I'm not 100% sure this change does not have implications beyond detecting duplicate args.

Likely the original author did not realize that Strings were valid arguments.

jez reviewed (changes requested):
whoops, forgot to request changes

This does not have any problems on Stripe's codebase, so if you can take care of these PR comments we can merge it.

EricHasegawa commented:
> whoops, forgot to request changes
> 
> This does not have any problems on Stripe's codebase, so if you can take care of these PR comments we can merge it.

Thanks! The code changed a decent amount after addressing comments, it's likely worth re-running on the codebase with the new changes imo

jez commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Rt9tcwSo2Oo6mc
→ https://go/builds/bui_Rt9tAxPhSfsEDP
→ https://go/builds/bui_Rt9ullkshLJDr8

jez commented:
I pushed up some changes to your branch.

This will merge when the tests pass.
